### DISCUSSION TITLE
Hello author, your research is very interesting. I am currently researching cable-driven robots. Does this toolbox support it?

### DISCUSSION THREAD

**Discussion by CJ0510** [BJCaasenbrood/SorotokiCode#14] [Q&A] [answered]
Hello author, your research is very interesting. I am currently researching cable-driven robots. Does this toolbox support it?

**Comment by BJCaasenbrood** [comment, marked as answer]
Yes but it involves some work. To implement cable-driven actuation, you are required to know how to compute the force vector/wrench. If I am not wrong, cable actuation can be written in the form: $\tau = J^\top(q) Ad_{\bar{g}}^\top u$ with $u = [m_1,m_2,m_3,n_1,n_2,n_3]^\top$ denoting the $\tau$ as the generalized forces, $J$ the Jacobian matrix that related joint-velocities to body velocities, and $u$ the input forces and moments denoted by $n_i$ and $m_i$, respectively; and $Ad_{\bar{g}}$ the adjoint transformation from backbone to the location/direction where the cable attaches itself to the body. For simplicity, we write $\tau = G(q) u$ where $G(q)$ is the input mapping. Note that the complexity here lies in computing $G(q)$ and thus $J(q)$.

Luckily, Sorotoki does compute the Jacobian for you once you made a `Shapes.m` class model of your soft robot. When calling `[g, J] = Shapes.string(q)` the Jacobian matrix is computed (along side the backbone curve $g \in \textrm{SE}(3)$). The matrix `J` is $6 \times N_q \times N_d$ with $N_q = \dim(q) $ and $N_d$ the nodal discretization. This means, it computes the Jacobian for every point on the backbone curve and you need to select the one you desired based on the spatial stepsize.

To give a quick example, here is how I would implement such a scheme:

```matlab
shp = Shapes(...)

% set shape class input mapping function
shp = shp.setInputMap(CableFnc);

% define the cable-actuation function
function G = CableFnc(shp)

   % get current joint position
   q = shp.solver.sol.x;
   
   % compute backbone g and Jacobians
   [g,J] = shp.string(q);
   
   id  = 20; % select node 20 (depends on where cable grasps)
   Ja1 = J(:,:,id);
   Ja2 = Admap(g(1:3,1:3,ii), [0;0;0]).' * Ja1 % transform Jacobian from global to local
   Ja3 = Admap(eye(3), [0;0;5]) * Ja2          % set cable location 5 mm away from backbone
   
   G = (Ja3(4,:).');                           % select the linear force n_1 as force director, 
                                               % i.e., 4th entry of 6-dim wrench vector [m,n].
end
```

I hope this answer helps.
